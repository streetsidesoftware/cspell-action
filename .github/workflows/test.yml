name: 'build-test'
on: # rebuild any PRs and main branch changes
  pull_request:
  push:
    branches:
      - main

permissions:
  contents: read

jobs:
  action-is-pristine: # make sure nothing changes in the action.
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v5
      - uses: streetsidesoftware/actions/public/setup-node-pnpm@v1

      - run: pnpm i
      - run: git --no-pager diff --compact-summary --exit-code

  clean-build: # make sure nothing changes with a clean build
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v5
      - uses: streetsidesoftware/actions/public/setup-node-pnpm@v1

      - run: pnpm i

      # Ensure the repository is clean after build
      - run: pnpm clean-build

      # test TypeScript build
      - run: |
          cd action-src
          pnpm build:tsc

      - run: git --no-pager diff --compact-summary --exit-code

  coverage: # run coverage and unit tests
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v5
        with:
          fetch-depth: 0

      - uses: streetsidesoftware/actions/public/setup-node-pnpm@v1
      - run: pnpm i

      - env:
          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
        run: GITHUB_OUTPUT="" pnpm coverage

      - name: Upload coverage Coveralls
        uses: coverallsapp/github-action@648a8eb78e6d50909eff900e4ec85cab4524a45b # 2.3.6
        with:
          github-token: ${{ secrets.GITHUB_TOKEN }}
          path-to-lcov: ./action-src/coverage/lcov.info

      - name: Upload coverage to Codecov
        uses: codecov/codecov-action@5a1091511ad55cbe89839c7260b706298ca349f7 # v3
        with:
          token: ${{ secrets.CODECOV_TOKEN }}
          directory: './action-src'
          files: ./coverage/lcov.info
