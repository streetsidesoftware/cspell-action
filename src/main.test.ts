import * as process from 'process';
import * as cp from 'child_process';
import * as path from 'path';
import * as helper from './test/helper';

const root = helper.root;
const mainFile = path.join(helper.debugDir, 'test_main.js');

const timeout = 20000;

describe('Validate Main', () => {
    test('GITHUB_TOKEN', () => {
        expect(process.env['GITHUB_TOKEN']).not.toBeUndefined();
    });

    test.each`
        test                                       | file
        ${'event pull_request main.js'}            | ${'pull_request.json'}
        ${'event pull_request_with_files main.js'} | ${'pull_request_with_files.json'}
        ${'event push main.js'}                    | ${'push.json'}
    `(
        '$test',
        ({ file }) => {
            const env = helper.fetchGithubActionFixture(file);
            env.FIXTURE_FILE_NAME = file;
            const options: cp.ExecSyncOptions = {
                env,
                cwd: root,
            };
            const result = cp.execSync(`node ${mainFile}`, options).toString();

            expect(cleanResult(result)).toMatchSnapshot();
        },
        timeout
    );
});

function cleanResult(output: string): string {
    // Remove time
    output = output.replace(/\(\d+\.\d\dms\)$/gm, '(???.??ms)');

    const rootRegExp = new RegExp(escapeRegEx(root), 'g');
    output = output.replace(rootRegExp, '.');

    return output;
}

function escapeRegEx(s: string) {
    return s.replace(/[-/\\^$*+?.()|[\]{}]/g, '\\$&');
}
