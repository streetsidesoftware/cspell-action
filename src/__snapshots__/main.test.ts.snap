// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Validate Main event pull_request main.js 1`] = `
"cspell-action
Pull Request
1/8 ./.github/workflows/test.yml (???.??ms)
2/8 ./.eslintrc.js (???.??ms)
3/8 ./package.json (???.??ms)
4/8 ./.vscode/launch.json (???.??ms)
5/8 ./fixtures/sampleCode/ts/sample.ts (???.??ms)
6/8 ./action.yaml (???.??ms)
7/8 ./README.md (???.??ms)
8/8 ./.github/workflows/lint.yml (???.??ms)
Files checked: 8, Issues found: 0 in 0 files.
Done.
"
`;

exports[`Validate Main event pull_request_with_files main.js 1`] = `
"cspell-action
Pull Request
1/5 ./.github/workflows/test.yml (???.??ms)
2/5 ./.eslintrc.js (???.??ms)
3/5 ./fixtures/sampleCode/ts/sample.ts (???.??ms)
4/5 ./action.yaml (???.??ms)
5/5 ./.github/workflows/lint.yml (???.??ms)
Files checked: 5, Issues found: 0 in 0 files.
Done.
"
`;

exports[`Validate Main event push main.js 1`] = `
"cspell-action
Push
1/1 ./.editorconfig (???.??ms)
Files checked: 1, Issues found: 0 in 0 files.
Done.
"
`;
